File: frontend/src/app/service/horse.service.ts
import {HttpClient} from '@angular/common/http';
import {Injectable} from '@angular/core';
import {Observable} from 'rxjs';
import {environment} from 'src/environments/environment';
import {Horse} from '../dto/horse';
import {HorseSearch} from "../dto/horse-search";

const baseUri = environment.backendUrl + '/horses';

@Injectable({
  providedIn: 'root'
})
export class HorseService {

  constructor(private http: HttpClient) {
  }

  /**
   * Get all horses with given search properties (name,description,birthdate,sex,ownerName)
   * @return returns list of found horses
   */
  searchHorses(horseSearchDto?: HorseSearch): Observable<Horse[]> {
    const searchParams: URLSearchParams = new URLSearchParams();
    if (horseSearchDto?.name) searchParams.append("name", horseSearchDto.name);
    if (horseSearchDto?.description) searchParams.append("description", horseSearchDto.description);
    if (horseSearchDto?.birthdate) searchParams.append("birthdate", horseSearchDto.birthdate.toString());
    if (horseSearchDto?.sex) searchParams.append("sex", horseSearchDto.sex);
    if (horseSearchDto?.ownerName) searchParams.append("ownerName", horseSearchDto.ownerName);
    return this.http.get<Horse[]>(`${baseUri}?${searchParams.toString()}`);
  }

  /**
   * Get horse by id
   * @return returns horse with id
   */
  getHorse(id: number): Observable<Horse> {
    return this.http.get<Horse>(`${baseUri}/${id}`);
  }


  /**
   *  Returns a horse with all corresponding horses
   * @param id must exist, gens has to be > 0
   * @return returns an optional horse and goes throw one horse recursively to get all corresponding horses
   */
  getHorseWithGenerations(id: number, gens: number): Observable<Horse> {
    return this.http.get<Horse>(`${baseUri}/${id}/familytree?generations=${gens}`);
  }

  /**
   * Get 5 closest matches where birthdate is after all others, the horseName which represents the input is Like the
   * horse names in the database and the gender in the database matches the given gender
   * @param birthdate, horseName, gender has to be valid
   * @return if you pass the wrong arguments it will throw a HttpClientErrorException.NotFound
   */
  getHorsesAutocomplete(birthdate: Date, horseName: string, gender: string): Observable<Horse[]> {
    return this.http.get<Horse[]>(`${baseUri}/autocomplete?birthdate=${birthdate}&horseName=${horseName}&gender=${gender}`);
  }

  /**
   * Deletes horse via id
   * @params id has to be valid
   */
  deleteHorse(id: number): Observable<object> {
    return this.http.delete(`${baseUri}/${id}`);
  }

  /**
   * Updates horse
   * @params the name, the birthdate and sex of horse has to be set the rest can be null
   */
  editHorse(horse: Horse): Observable<object> {
    return this.http.put(baseUri, horse);
  }

  /**
   * Inserts horse
   * @params the name, the birthdate and sex of horse has to be set the rest can be null
   */
  insertHorse(horse: Horse): Observable<object> {
    return this.http.post(baseUri, horse);
  }
}

File: backend/src/main/java/at/ac/tuwien/sepm/assignment/individual/service/HorseService.java
package at.ac.tuwien.sepm.assignment.individual.service;

import at.ac.tuwien.sepm.assignment.individual.dto.HorseDto;
import at.ac.tuwien.sepm.assignment.individual.dto.HorseSearchDto;
import at.ac.tuwien.sepm.assignment.individual.entity.Horse;

import java.time.LocalDate;
import java.util.List;
import java.util.Optional;

public interface HorseService {
    /**
     * Get all horses with given search properties (name,description,birthdate,sex,ownerName)
     * @return returns list of found horses
     * @throws if there are no horses it will throw a HttpClientErrorException.NotFound
     */
    List<Horse> searchHorses(HorseSearchDto horseSearchDto);

    /**
     * Get horse by id
     * @return returns horse with id
     * @throws if there is a problem throw a HttpClientErrorException.NotFound
     */
    Optional<Horse> getHorse(Long id);

    /**
     * Inserts horse
     * @params the name, the birthdate and sex of h has to be set the rest can be null
     * @throws if you pass the wrong arguments it will throw a HttpClientErrorException.NotFound
     */
    void insertHorse(HorseDto h);

    /**
     * Updates horse
     * @params the name, the birthdate and sex of h has to be set the rest can be null
     * @throws if you pass the wrong arguments it will throw a HttpClientErrorException.NotFound
     */
    void updateHorse(HorseDto h);

    /**
     * Deletes horse via id
     * @params id has to be valid
     * @throws if you pass the wrong arguments it will throw a HttpClientErrorException.NotFound
     */
    void deleteHorse(Long id);

    /**
     *  Returns a horse with all corresponding horses
     * @param id must exist, gens has to be > 0
     * @return returns an optional horse and goes throw one horse recursively to get all corresponding horses
     */
    Optional<Horse> getHorseWithGenerations(Long id, int gens);

    /**
     * Get 5 closest matches where birthdate is after all others, the horseName which represents the input is Like the
     * horse names in the database and the gender in the database matches the given gender
     * @param birthdate, horseName, gender has to be valid
     * @return if you pass the wrong arguments it will throw a HttpClientErrorException.NotFound
     */
    List<Horse> getHorsesAutoComplete(LocalDate birthdate, String horseName, String gender);
}
